Mage.Sets/src/mage/cards/t/TorrentOfLava.java

```diff
@@ -0,0 +1,128 @@
+package mage.cards.t;
+
+import mage.abilities.Ability;
+import mage.abilities.common.SimpleActivatedAbility;
+import mage.abilities.common.SimpleStaticAbility;
+import mage.abilities.costs.common.TapSourceCost;
+import mage.abilities.dynamicvalue.common.ManacostVariableValue;
+import mage.abilities.effects.common.DamageAllEffect;
+import mage.abilities.effects.common.PreventDamageToSourceEffect;
+import mage.abilities.effects.common.continuous.GainAbilityAllEffect;
+import mage.abilities.keyword.FlyingAbility;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.Zone;
+import mage.filter.StaticFilters;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.filter.predicate.Predicates;
+import mage.filter.predicate.mageobject.AbilityPredicate;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.stack.Spell;
+
+import java.util.UUID;
+
+/**
+ * @author Cguy7777
+ */
+public final class TorrentOfLava extends CardImpl {
+
+    private static final FilterCreaturePermanent filter = new FilterCreaturePermanent("creature without flying");
+
+    static {
+        filter.add(Predicates.not(new AbilityPredicate(FlyingAbility.class)));
+    }
+
+    public TorrentOfLava(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{X}{R}{R}");
+
+        // Torrent of Lava deals X damage to each creature without flying.
+        this.getSpellAbility().addEffect(new DamageAllEffect(ManacostVariableValue.REGULAR, filter));
+
+        // As long as Torrent of Lava is on the stack, each creature has
+        // "{tap}: Prevent the next 1 damage that would be dealt to this creature by Torrent of Lava this turn."
+        this.addAbility(new SimpleStaticAbility(Zone.STACK, new TorrentOfLavaGainAbilityEffect()));
+    }
+
+    private TorrentOfLava(final TorrentOfLava card) {
+        super(card);
+    }
+
+    @Override
+    public TorrentOfLava copy() {
+        return new TorrentOfLava(this);
+    }
+}
+
+class TorrentOfLavaGainAbilityEffect extends GainAbilityAllEffect {
+
+    TorrentOfLavaGainAbilityEffect() {
+        super(new SimpleActivatedAbility(
+                        new TorrentOfLavaPreventionEffect(null, 0, null), new TapSourceCost()),
+                Duration.Custom,
+                StaticFilters.FILTER_PERMANENT_CREATURES);
+        this.staticText = "As long as {this} is on the stack, " +
```

Comment:
Use dynamic getrules with additional text at the end like CardUtils.getSourceXXX logs text.

Users must see card id, so it can help to identify lava instance. If doesn't work then add static card hint to the prevent ability with same info.

Reply:
See comment below about dynamic card hint - its better solution for a users.